Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,242 @@
+        )
+
+        context.write(self.wrapOutboundOut(.head(head)), promise: nil)
+        context.write(self.wrapOutboundOut(.end(nil)), promise: nil)
+        context.flush()
+    }
+
+    private func handleHTTPHeadReceived(_ head: HTTPResponseHead, context: ChannelHandlerContext) {
+        switch self.state {
+        case .connectSent(let scheduled):
+            switch head.status.code {
+            case 200..<300:
+                // Any 2xx (Successful) response indicates that the sender (and all
+                // inbound proxies) will switch to tunnel mode immediately after the
+                // blank line that concludes the successful response's header section
+                self.state = .headReceived(scheduled)
+            case 407:
+                self.failWithError(Error.proxyAuthenticationRequired, context: context)
+
+            default:
+                // Any response other than a successful response indicates that the tunnel
+                // has not yet been formed and that the connection remains governed by HTTP.
+                self.failWithError(Error.invalidProxyResponseHead(head), context: context)
+            }
+        case .failed:
+            break
+        case .initialized, .headReceived, .completed:
+            preconditionFailure("Invalid state: \(self.state)")
+        }
+    }
+
+    private func handleHTTPBodyReceived(context: ChannelHandlerContext) {
+        switch self.state {
+        case .headReceived(let timeout):
+            timeout.cancel()
+            // we don't expect a body
+            self.failWithError(Error.invalidProxyResponse, context: context)
+        case .failed:
+            // ran into an error before... ignore this one
+            break
+        case .completed, .connectSent, .initialized:
+            preconditionFailure("Invalid state: \(self.state)")
+        }
+    }
+
+    private func handleHTTPEndReceived(context: ChannelHandlerContext) {
+        switch self.state {
+        case .headReceived(let timeout):
+            timeout.cancel()
+            self.state = .completed
+            self.promise.succeed(())
+
+        case .failed:
+            // ran into an error before... ignore this one
+            break
+        case .initialized, .connectSent, .completed:
+            preconditionFailure("Invalid state: \(self.state)")
+        }
+    }
+
+    private func failWithError(_ error: Error, context: ChannelHandlerContext, closeConnection: Bool = true) {
+        self.state = .failed(error)
+        self.promise.fail(error)
+        context.fireErrorCaught(error)
+        if closeConnection {
+            context.close(mode: .all, promise: nil)
+        }
+    }
+
+    /// Error types for ``HTTP1ProxyConnectHandler``
+    public struct Error: Swift.Error, Equatable {
+
+        fileprivate enum ErrorEnum: Equatable {
+            case proxyAuthenticationRequired
+            case invalidProxyResponseHead(head: HTTPResponseHead)
```

Comment:
I think we may want to add a custom `Equatable` implementation and ignore the `HTTPResponseHead` in the comparison. Therefore two `invalidProxyResponseHead` errors are equal, regardless of the exact `head`. We have already done something similar if we include the file and line number where we ignore them in the `Equatable` and `Hashable` implementation. @Lukasa what do you think about that?

In addition, `Hashable` conformance would be nice as well.

Reply:
Yes, that seems wise.